pkg/policy/api/egress.go

```diff
@@ -148,12 +148,37 @@ type EgressRule struct {
 	//     - 'sg-XXXXXXXXXXXXX'
 	// +optional
 	ToGroups []ToGroups `json:"toGroups,omitempty"`
+
+	aggregatedSelectors EndpointSelectorSlice
+}
+
+// SetAggregatedSelectors creates a single slice containing all of the following
+// fields within the EgressRule, converted to EndpointSelector, to be stored
+// within the EgressRule for easy lookup while performing policy evaluation
+// for the rule:
+// * ToEntities
+// * ToCIDR
+// * ToCIDRSet
+// * ToFQDNs
+//
+// ToEndpoints is not aggregated due to requirement folding in
+// GetDestinationEndpointSelectorsWithRequirements()
+func (e *EgressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
+	res = append(res, e.ToEntities.GetAsEndpointSelectors()...)
+	res = append(res, e.ToCIDR.GetAsEndpointSelectors()...)
+	res = append(res, e.ToCIDRSet.GetAsEndpointSelectors()...)
+	res = append(res, e.ToFQDNs.GetAsEndpointSelectors()...)
+	e.aggregatedSelectors = res
 }
 
 // GetDestinationEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
 // all L3 source selectors of the ingress rule
 func (e *EgressRule) GetDestinationEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
-	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
+	if e.aggregatedSelectors == nil {
+		e.SetAggregatedSelectors()
```

Comment:
There is no protection against parallel usage of `EgressRule` for this to happen safely.

Reply:
It's safe even if multiple goroutines run this in parallel. They will compute the same result so the order in which they set the pointer does not matter. I've added a comment to this effect.

Reply:
this feels really odd... also, @jrajahalme comment not added :-)

Reply:
The comment was added in the implementation of `SetAggregatedSelectors`.

Reply:
The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)